ecommerce/api.py

```diff
@@ -0,0 +1,105 @@
+"""
+Functions for ecommerce
+"""
+from base64 import b64encode
+import hashlib
+import hmac
+import logging
+import uuid
+
+from django.conf import settings
+
+from mitxpro.utils import now_in_utc
+
+ISO_8601_FORMAT = "%Y-%m-%dT%H:%M:%SZ"
+log = logging.getLogger(__name__)
+
+_REFERENCE_NUMBER_PREFIX = "MITXPRO-"
+
+
+def generate_cybersource_sa_signature(payload):
+    """
+    Generate an HMAC SHA256 signature for the CyberSource Secure Acceptance payload
+    Args:
+        payload (dict): The payload to be sent to CyberSource
+    Returns:
+        str: The signature
+    """
+    # This is documented in certain CyberSource sample applications:
+    # http://apps.cybersource.com/library/documentation/dev_guides/Secure_Acceptance_SOP/html/wwhelp/wwhimpl/js/html/wwhelp.htm#href=creating_profile.05.6.html
+    keys = payload["signed_field_names"].split(",")
+    message = ",".join(f"{key}={payload[key]}" for key in keys)
+
+    digest = hmac.new(
+        settings.CYBERSOURCE_SECURITY_KEY.encode("utf-8"),
+        msg=message.encode("utf-8"),
+        digestmod=hashlib.sha256,
+    ).digest()
+
+    return b64encode(digest).decode("utf-8")
+
+
+def generate_cybersource_sa_payload(order):
+    """
+    Generates a payload dict to send to CyberSource for Secure Acceptance
+    Args:
+        order (Order): An order
+    Returns:
+        dict: the payload to send to CyberSource via Secure Acceptance
+    """
+    # http://apps.cybersource.com/library/documentation/dev_guides/Secure_Acceptance_WM/Secure_Acceptance_WM.pdf
+    # Section: API Fields
+
+    # NOTE: be careful about max length here, many (all?) string fields have a max
+    # length of 255. At the moment none of these fields should go over that, due to database
+    # constraints or other reasons
```

Comment:
Does CyberSource provide a meaningful error back to us if this is the case or should we validate this constraint and raise/log an error before returning here to ensure we realize it sooner?

Reply:
I don't know, I think I added this in MicroMasters because I saw the limitation mentioned in the docs. I don't think we've ever run into it. I'll add a `[:255]` to truncate the description and other string fields.

Reply:
Oh interesting. I'm pretty sure we broke that at least once in Bootcamps. I'll see if I can find the error in sentry...

Reply:
ok, truncation is ok by me if it's just an arbitrary value and not functional

Reply:
Oh interesting... maybe we should double check this function in bootcamp and micromasters to make sure the string fields are capped